core/src/connection/pool.rs

```diff
@@ -0,0 +1,907 @@
+use std::{error, fmt, hash::Hash, task::Context, task::Poll};
+
+/// A connection `Pool` manages a set of connections for each peer.
+pub struct Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo = PeerId, TPeerId = PeerId> {
+    local_id: TPeerId,
+
+    /// The configuration of the pool.
+    limits: PoolLimits,
+
+    /// The connection manager that handles the connection I/O for both
+    /// established and pending connections.
+    ///
+    /// For every established connection there is a corresponding entry in `established`.
+    manager: Manager<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo>, // tasks
+
+    /// The managed connections of each peer that are currently considered
+    /// established, as witnessed by the associated `ConnectedPoint`.
+    established: FnvHashMap<TPeerId, FnvHashMap<ConnectionId, ConnectedPoint>>,
+
+    /// The pending connections that are currently being negotiated.
+    pending: FnvHashMap<ConnectionId, (ConnectedPoint, Option<TPeerId>)>,
+}
+
+impl<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> fmt::Debug
+for Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId>
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        // TODO: More useful debug impl?
+        f.debug_struct("Pool")
+            .field("limits", &self.limits)
+            .finish()
+    }
+}
+
+impl<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> Unpin
+for Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> {}
+
+/// Event that can happen on the `Pool`.
+pub enum PoolEvent<'a, TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId> {
+    /// A new connection has been established.
+    ///
+    /// In order for the connection to remain in the pool, it must be accepted.
+    /// If the `NewConnection` is dropped without being accepted, the connection
+    /// is closed and removed from the pool.
+    ConnectionEstablished {
+        connection: EstablishedConnection<'a, TInEvent, TConnInfo, TPeerId>,
+        num_established: usize,
+    },
+
+    /// A newly established connection has been dropped because of the configured
+    /// per-peer connection limit.
+    ConnectionLimitReached {
+        connected: Connected<TConnInfo>,
+        info: ConnectionLimit,
+    },
+
+    /// An established connection has encountered an error.
+    ///
+    /// Can only happen after a node has been successfully reached.
+    ConnectionError {
+        id: ConnectionId,
+        /// Information about the connection that errored.
+        connected: Connected<TConnInfo>,
+        /// The error that happened.
+        error: ConnectionError<THandlerErr, TTransErr>,
+        /// A reference to the pool that used to manage the connection.
+        pool: &'a mut Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo, TPeerId>,
```

Comment:
We normally don't need these fields, but I also see why you added them. I'm fine with merging this PR then seeing afterwards if we can satisfy the borrow checker without making the code significantly dirtier.